### PR TITLE
Data-driven file upload

### PR DESCRIPTION
**Pull request by igorkravchenko**
This PR is aimed to provide flow for selection of file attachment for message on Message Center Welcome screen and reworked version of FileUploadListView that uses data-driven approach. Currently FileUploadListView is used by ChatView, but in subsequent PR(s) it will be used by WelcomeView as well.

MOB-1815

### CODE REVIEW

File: GliaWidgets/SecureConversations/SecureConversations.Coordinator.swift [hunk 1]
```diff
@@ -39,7 +53,7 @@ extension SecureConversations {
                 props: viewModel.props()
             )
 
-            viewModel.delegate = { [weak self, weak controller] event in
+            viewModel.delegate = { [weak self, weak controller, style = viewFactory.theme.secureConversationsWelcomeStyle] event in
```
Comment (dukhovnyi) [inline GliaWidgets/SecureConversations/SecureConversations.Coordinator.swift:56]: will style be retained here (strong reference)?

File: GliaWidgets/SecureConversations/SecureConversations.Coordinator.swift [hunk 2]
```diff
@@ -90,13 +144,54 @@ extension SecureConversations {
...
+            observable.addObserver(self, update: { newValue, _ in pickerEvent(newValue) })
+            let viewModel = MediaPickerViewModel(
+                pickerEvent: observable,
+                mediaSource: mediaSource,
+                mediaTypes: mediaTypes
+            )
+            let controller = MediaPickerController(viewModel: viewModel)
+            controller.viewController { [weak self] viewController in
+                self?.navigationPresenter.present(viewController)
+            }
```
Comment (dukhovnyi) [inline GliaWidgets/SecureConversations/SecureConversations.Coordinator.swift:163]: a little bit complicated and not a straightforward construction ...

File: GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift [hunk 3]
```diff
@@ -0,0 +1,965 @@
...
+  import QuickLookThumbnailing
+#endif
+
+private extension Int {
+    static let maxUnscrollableViewsOnDefaultContentSizeCategory = 3
+    static let maxUnscrollableViewsOnLargeContentSizeCategory = 2
+}
+
+extension SecureConversations {
+    class FileUploadListView: UIView {
```
Comment (dukhovnyi) [inline GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift:12]: I think we can inherit from `BaseView` and use `setup` & `defineLayout` functions, does it make sense?
```suggestion
    final class FileUploadListView: BaseView {
```

File: GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift [hunk 4]
```diff
@@ -0,0 +1,965 @@
...
+        private func setup() {
+            stackView.axis = .vertical
+            // Assign empty array, because `accessibilityElements` is nil initially,
+            // and we need to append to/remove from it when views
+            // are added to/removed from stack view
+            stackView.accessibilityElements = []
+        }
+
+        private func layout() {
+            heightLayoutConstraint = autoSetDimension(.height, toSize: 0)
```
Comment (dukhovnyi) [inline GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift:78]: let's minimize using PureLayout and start using native approach or our own helper-functions.

File: GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift [hunk 5]
```diff
@@ -0,0 +1,965 @@
...
+            // Assign props for corresponding views
+            for uploadProps in props.uploads {
+                cachedViews[by: uploadProps.id]?.props = uploadProps
+            }
+        }
+    }
+}
+
+extension SecureConversations.FileUploadListView {
+    class FileUploadView: UIView {
```
Comment (dukhovnyi) [inline GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift:160]: I'd like to propose
- make this class final
- inherit from BaseView
- move to separate file

File: GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift [hunk 6]
```diff
@@ -0,0 +1,965 @@
...
+
+        private let contentView = UIView()
+        private let infoLabel = UILabel()
+        private let stateLabel = UILabel()
+        private let filePreviewView = FilePreviewView()
+        private let progressView = UIProgressView()
+        let removeButton = UIButton()
+
+        private let kContentInsets = UIEdgeInsets(top: 10, left: 10, bottom: 10, right: 10)
+        private let kRemoveButtonSize = CGSize(width: 30, height: 30)
```
Comment (dukhovnyi) [inline GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift:181]: I understand that these are legacy, however, if you have chance, please drop `k` prefix for these variables.

File: GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift [hunk 7]
```diff
@@ -0,0 +1,965 @@
...
+
+                // NO LONGER NEEDED
+//                progress.addObserver(self) { [weak self, accessibility = style.accessibility] progress, _ in
+//                    self?.progressView.progress = Float(progress)
+//                    self?.accessibilityValue = Self.accessibleProgress(
+//                        provideProgressText(progress),
+//                        to: self?.infoLabel.text,
+//                        accessibility: accessibility
+//                    )
+//                }
```
Comment (dukhovnyi) [inline GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift:303]: remove?

File: GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift [hunk 8]
```diff
@@ -0,0 +1,965 @@
...
+                }
+            default:
+                self = .generic
+            }
+        }
+    }
+}
+
+extension SecureConversations.FileUploadListView.FileUploadView {
+    class FilePreviewView: UIView {
```
Comment (dukhovnyi) [inline GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift:430]: - does it make sense to move it to separate file?
- does it make sense to make it as final?
- does it make sense to inherit from BaseView?

**Review by dukhovnyi — APPROVED** [review]

**Review by rasmustautsglia — APPROVED** [review]

File: GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift [hunk 9]
```diff
@@ -0,0 +1,965 @@
...
+        @available(*, unavailable)
+        required init?(coder: NSCoder) {
+            fatalError("init(coder:) has not been implemented")
+        }
+
+        private func setup() {
+            clipsToBounds = true
+
+            label.textAlignment = .center
+
```
Comment (EgorovEI) [inline GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift:470]: unneeded space

File: GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift [hunk 10]
```diff
@@ -0,0 +1,965 @@
...
+}
+
+private extension FileUploadStyle {
+    static let initial = Theme().chat.messageEntry.uploadList.item
+}
+
+// TODO: Add unit tests
+// MOB-1815
+/// Collection that allows easier access by element identifier.
+struct IdCollection<Identifier: Hashable, Item: Equatable>: Equatable & Collection {
```
Comment (EgorovEI) [inline GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift:636]: Would it be reasonable to move it to separate file?

File: GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift [hunk 11]
```diff
@@ -0,0 +1,965 @@
...
+
+        func onUploaderStateChanged(_ state: FileUploader.State) {
+            reportChange()
+        }
+
+        func removeUpload(_ upload: FileUpload) {
+            upload.state.removeObserver(self)
+            environment.uploader.removeUpload(upload)
+            reportChange()
+//            action?(.removeUpload(upload))
```
Comment (EgorovEI) [inline GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift:804]: Will it be uncommented or can be deleted?

File: GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift [hunk 12]
```diff
@@ -0,0 +1,965 @@
...
+                uploads: uploads
+            )
+            return props
+        }
+    }
+}
+
+// MARK: - Compatibility extension for ChatViewModel
+extension SecureConversations.FileUploadListViewModel {
+    var  attachment: CoreSdkClient.Attachment? {
```
Comment (EgorovEI) [inline GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift:836]:
```suggestion
    var attachment: CoreSdkClient.Attachment? {
```

File: GliaWidgets/SecureConversations/Welcome/SecureConversations.WelcomeViewModel.swift [hunk 13]
```diff
@@ -197,5 +231,152 @@ extension SecureConversations.WelcomeViewModel {
...
+        self.uploader = uploader
+    }
+}
+
+// MARK: Support alerts system
+extension SecureConversations.WelcomeViewModel {
+    private static var alertPresenters = Set<SecureConversations.WelcomeViewModel>()
+}
+
+extension  SecureConversations.WelcomeViewModel: Hashable {
```
Comment (EgorovEI) [inline GliaWidgets/SecureConversations/Welcome/SecureConversations.WelcomeViewModel.swift:374]:
```suggestion
extension SecureConversations.WelcomeViewModel: Hashable {
```

File: GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift [hunk 14]
```diff
@@ -0,0 +1,965 @@
...
+        static let height: CGFloat = 60
+
+        private let contentView = UIView()
+        private let infoLabel = UILabel()
+        private let stateLabel = UILabel()
+        private let filePreviewView = FilePreviewView()
+        private let progressView = UIProgressView()
+        let removeButton = UIButton()
+
+        private let kContentInsets = UIEdgeInsets(top: 10, left: 10, bottom: 10, right: 10)
```
[the remark this answers is not in the document]
Comment (igorkravchenko) [reply GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift:180]: This is a copy of old (non-data driven) view. The idea is to make it data-driven with minimal effort, so I do not touch any unrelated code for data-driven approach. This has to be changed at some point, but that is not the goal of this PR.

File: GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift [hunk 15]
```diff
@@ -0,0 +1,965 @@
...
+        private func setup() {
+            infoLabel.lineBreakMode = .byTruncatingMiddle
+            stateLabel.adjustsFontSizeToFitWidth = true
+
+            progressView.clipsToBounds = true
+            progressView.layer.cornerRadius = 4
+
+            removeButton.addTarget(self, action: #selector(remove), for: .touchUpInside)
+            isAccessibilityElement = true
+
```
[the remark this answers is not in the document]
Comment (igorkravchenko) [reply GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift]: Fixed 77e44179c0de4b21cc95d9a76b5dc8bdb34eaa33.

File: GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift [hunk 16]
```diff
@@ -0,0 +1,965 @@
...
+        // swiftlint:disable function_body_length
+        private func update(for state: FileUpload.State) {
+            switch state {
+            case .none:
+                filePreviewView.props = .init(style: props.style.filePreview, kind: .none)
+                infoLabel.text = nil
+                stateLabel.text = nil
+                accessibilityValue = nil
+            case  let.uploading(progress, localFile):
```
[the remark this answers is not in the document]
Comment (igorkravchenko) [reply GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift]: Fixed 1867d4a50a32dd5474f74beb42c12cc3fdcad03a.

File: GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift (hunk 9 above)
Comment (igorkravchenko) [reply GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift]: Fixed 5c3177cf83710b930a85beea32af44ac027ffd14.

File: GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift (hunk 11 above)
Comment (igorkravchenko) [reply GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift]: Removed 3ecc2f54eca61a1d236e51e3048b292b4ef28583.

File: GliaWidgets/SecureConversations/Welcome/SecureConversations.WelcomeViewModel.swift (hunk 13 above)
Comment (igorkravchenko) [reply GliaWidgets/SecureConversations/Welcome/SecureConversations.WelcomeViewModel.swift]: Removed spacing 3102cf12dd73c1ffb6e3dd52784d33c2e582a0e3.

File: GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift [hunk 17]
```diff
@@ -0,0 +1,965 @@
...
+                progressView.progress = Float(progress)
+                let provideProgressText: (Double) -> String = { "\(Int($0 * 100))" }
+                accessibilityValue = Self.accessibleProgress(
+                    provideProgressText(progress),
+                    to: infoLabel.text,
+                    accessibility: props.style.accessibility
+                )
+
+                // NO LONGER NEEDED
+//                progress.addObserver(self) { [weak self, accessibility = style.accessibility] progress, _ in
```
[the remark this answers is not in the document]
Comment (igorkravchenko) [reply GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift]: Sure bf3e51f3060149a70ed4c1e37bfcc18f1420be36.

File: GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift (hunk 10 above)
Comment (igorkravchenko) [reply GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift]: Very reasonable 41274fac4e0a17be2e089747b4086c92b1c2934a.

**Review by EgorovEI — APPROVED** [review]

**Comment by igorkravchenko** [comment]
!squash

**Comment by igorkravchenko** [comment]
!merge

**Pull request merged**